packages/sdk/client-services/src/packlets/tests/text.test.ts

```diff
@@ -0,0 +1,210 @@
+      model.peers.has(this.peer.halo.identity!.identityKey) &&
+      model.text.length <= this.index &&
+      this.index >= this.length
+    );
+  }
+
+  async run(model: Model, real: Real) {
+    model.text = model.text.slice(0, this.index - length) + model.text.slice(this.index);
+
+    const space = this.peer.echo.getSpace(real.spaceKey)!;
+    const [document] = space.db.query((obj) => !!obj.content).objects;
+    const text = (document.content.doc as Doc).getText('utf8');
+    text.delete(this.index, this.length);
+
+    await assertState(model, real);
+  }
+
+  toString() {
+    const peer = this.peer.halo.identity?.identityKey.truncate();
+    return `RemoveText(peer=${peer}, index=${this.index}, length=${this.length})`;
+  }
+}
+
+describe.only('Client text replication', () => {
+  test('property-based tests', async () => {
+    const testBuilder = new TestBuilder();
+    const host = testBuilder.createClientServicesHost();
+    await host.open();
+    afterTest(() => host.close());
+    const [client, server] = testBuilder.createClientServer(host);
+    void server.open();
+    afterTest(() => server.close());
+    await client.initialize();
+    afterTest(() => client.destroy());
+    await client.halo.createIdentity();
+    const space = await client.echo.createSpace();
+    await space.db.add(new Document({ content: new Text() }));
+
+    const peers = await Promise.all(
+      range(10).map(async () => {
+        const host = testBuilder.createClientServicesHost();
+        await host.open();
+        afterTest(() => host.close());
+        const [client, server] = testBuilder.createClientServer(host);
+        void server.open();
+        afterTest(() => server.close());
+        await client.initialize();
+        afterTest(() => client.destroy());
+        await client.halo.createIdentity();
+        return client;
+      })
+    );
+    const peer = fc.constantFrom(...peers);
+
+    const allCommands = [
+      peer.map((peer) => new CreatePeerCommand(client, peer)),
+      fc.tuple(peer, fc.integer(), fc.unicode()).map(([peer, index, text]) => new InsertTextCommand(peer, index, text)),
+      fc
+        .tuple(peer, fc.integer(), fc.integer())
+        .map(([peer, index, length]) => new RemoveTextCommand(peer, index, length))
+    ];
+
+    await fc.assert(
+      fc.asyncProperty(fc.commands(allCommands), async (commands) => {
+        const setup: ModelRunSetup<Model, Real> = () => {
+          const model = {
+            text: '',
+            peers: new ComplexSet<PublicKey>(PublicKey.hash)
+          };
+          const real = {
+            spaceKey: space.key,
+            peers: new ComplexMap<PublicKey, Client>(PublicKey.hash)
+          };
+
+          model.peers.add(client.halo.identity!.identityKey);
+          real.peers.set(client.halo.identity!.identityKey, client);
+
+          return { model, real };
+        };
+
+        await fc.asyncModelRun(setup, commands);
+      }),
+      {
+        examples: [
```

Comment:
can you explain?
is this a starting point for more randomized tests?

Reply:
yeah, exactly, I'm expecting we'll fill out more examples once it's working but just trying to get a very simple one passing first to make sure it's structured right as I haven't used this before and am just modelling off of @dmaretskyi 's network manager tests